File: lib/product/manifest_compatibility_v2.dart
final class ManifestCompatibilityV2 {
  const ManifestCompatibilityV2();

  String classify(Map<String, Object?> envelope) {
    final version = envelope['schemaVersion']?.toString() ?? '';
    if (const <String>{'1', '1.0', '1.0.0', 'v1'}.contains(version)) {
      throw const FormatException('v1_trust_disabled');
    }
    if (version != '2.0.0') {
      throw const FormatException('unsupported_manifest_version');
    }
    for (final field in const <String>{
      'hmac',
      'secret',
      'keyMaterial',
      'signingKey',
      'algorithm',
    }) {
      if (envelope.containsKey(field)) {
        throw const FormatException('mixed_format_rejected');
      }
    }
    return 'signed_manifest_v2';
  }
}

File: tool/manifest_compatibility_v2.py
#!/usr/bin/env python3
from __future__ import annotations

from typing import Any, Mapping


class ManifestCompatibilityError(ValueError):
    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code


LEGACY_FIELDS = frozenset({"hmac", "secret", "keyMaterial", "signingKey", "algorithm"})


def classify_manifest(value: Mapping[str, Any]) -> str:
    version = str(value.get("schemaVersion") or "")
    if version in {"1", "1.0", "1.0.0", "v1"}:
        raise ManifestCompatibilityError(
            "v1_trust_disabled",
            "Signed Manifest v1 can never authorize production trust.",
        )
    if version != "2.0.0":
        raise ManifestCompatibilityError(
            "unsupported_manifest_version",
            f"Unsupported manifest version: {version or '<missing>'}",
        )
    legacy = sorted(LEGACY_FIELDS.intersection(value))
    if legacy:
        raise ManifestCompatibilityError(
            "mixed_format_rejected",
            f"Signed Manifest v2 contains legacy trust fields: {legacy}",
        )
    return "signed_manifest_v2"
